### DISCUSSION TITLE
Mesh refinement with block vectors

### DISCUSSION THREAD

**Discussion by starkekr** [mfem/mfem#4160] [General]
Is there a proper way to refine a block vector? I have tried creating grid function views into the individual blocks using MakeTRef and calling the update method on each of the indiviual grid functions. This does not seem to update the original block vector though. I think I need a function that copies the original vector in blocked chunks, create a view on that chunk then use the functionality of the grid functions to interpolate the values on the resized chunks. Is there another way of doing this? I couldn't find an example that used both block vectors and mesh refinement.

**Comment by v-dobrev** [comment]
After refinement, and updating all `(Par)FiniteElementSpace`s, `(Par)GridFunction`s, `(Par)LinearForm`, and `(Par)BilinearForm` (e.g. similar to `UpdateAndRebalance()` in `ex15p.cpp`), you can propagate the changes to `BlockVector` by recomputing the block-offsets and calling `BlockVector::Update(const Array<int> &bOffsets)` with them. After that you will also need to re-associate your `(Par)GridFunction`s' internal true-vector with the respective blocks in the new `BlockVector` using `MakeTRef` as before. Finally, you can update the values inside the `BlockVector` by calling `(Par)GridFunction::SetTrueVector()` for each `(Par)GridFunction`. If you are using a code similar to `ex15`/`ex15p`, you can probably just add these `BlockVector`-related updates to `UpdateAndRebalance()`.

**Reply by starkekr** [reply]
That seemed to work, thanks!
